### ISSUE TITLE
Arithmetic operators for durations, dates, and times

### ISSUE THREAD

**Issue by jindrichmynarz** [w3c/sparql-12#32]
## Why?

Querying temporal data often requires arithmetic operations on durations, dates, and times. For example, we might want to retrieve books that were published during the past year. Such query currently either needs to hard-code the publication date of the oldest included book, or makes use of a non-standard behaviour or an extension function of a specific SPARQL engine.

SPARQL 1.1 [reuses arithmetic operators](https://www.w3.org/TR/sparql11-query/#OperatorMapping) for numerics from [XPath](https://www.w3.org/TR/xpath-functions/#numeric-functions). The behaviour of arithmetic operators, such as addition or subtraction, for durations, dates, and times is undefined.

## What could it look like?

Several SPARQL engines already allow some arithmetic operations on durations, dates, and times, such as Apache Jena, where the following query returns the date a year ago from now:

```rq
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (now() - "P1Y"^^xsd:duration AS ?year_ago)
WHERE {}
```

Similarly to arithmetic operators for numerics in SPARQL 1.1, SPARQL 1.2 can reuse the arithmetic operators for durations, dates, and times from [XPath](https://www.w3.org/TR/xpath-functions-31/#dateTime-arithmetic) by defining extending the [mapping of operators to XPath functions](https://www.w3.org/TR/sparql11-query/#OperatorMapping).

## Previous work

Some SPARQL engines offer extension functions for arithmetic on durations, dates, and times, such as `bif:datediff()` in OpenLink's Virtuoso.

## Considerations for backward compatibility

This change defines a previously undefined behaviour, so it is backwards-compatible.

**Comment by RickMoynihan** [comment]
Also I'd be tempted to add explicit support for `xsd:date` and `xsd:time` with comparisons; i.e. beyond just SPARQL 1.1's `xsd:dateTime`.

**Comment by jaw111** [comment]
Would it make sense to make it explicitly `xsd:yearMonthDuration` and `xsd:dayTimeDuration` instead of  `xsd:duration`?

Also suggest to enumerate the arithmetic operators that are proposed and which duration, date and time datatypes are allowed as arguments.

**Comment by afs** [comment]
See issue #55.

**Comment by jpcs** [comment]
+1 MarkLogic already supports full XPath arithmetic operations from SPARQL.

**Comment by namedgraph** [comment]
The example works in Dydra when I change the literal datatype to `xsd:yearMonthDuration`:
```sparql
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (now() - "P1Y"^^xsd:yearMonthDuration AS ?year_ago)
WHERE {}
```
Result:
```
?year_ago
--
2018-05-28T12:14:45Z
```
Also related: [Determining Day Of Week For Datetime in SPARQL](https://stackoverflow.com/questions/41729208/determining-day-of-week-for-datetime-in-sparql)
/cc @lisp

**Comment by kasei** [comment]
I've added some prototype tests for duration arithmetic in kasei@7c1b05e.

**Comment by JervenBolleman** [comment]
@kasei thank you! It is fantastic that you are picking this up!

**Comment by jaw111** [comment]
I suggest to extend this ticket to also cover comparison operators to support `ORDER BY` on durations, dates and times. Specifically mentioning the Gregorian datatypes `xs:gYearMonth`, `xs:gYear`, `xs:gMonthDay`, `xs:gMonth` and `xs:gDay` where only equality comparison operation is defined, but [section 9.4 of XPath and XQuery Functions and Operators 3.1](https://www.w3.org/TR/xpath-functions-3/#comp.datetime) how comparison can be done by calculating the starting instants.

Example query to count how many resources were created per year and month:

```sparql
prefix dct: <http://purl.org/dc/terms/>
select ?gYearMonth (count(*) as ?c)
where {
  ?s dct:created ?created .
  bind(xsd:gYearMonth(?created) as ?gYearMonth)
}
group by ?gYearMonth
order by ?gYearMonth
```

Obviously this could be done quite trivially using `year()` and `month()` functions, but one can imagine more complex cases.

**Comment by kasei** [comment]
@jaw111 –

I hadn't much considered adding `gYearMonth` or `gMonthDay` in my tests or implementation (or the broken out `gYear`, `gMonth`, or `gDay` types for that matter). Do you have use cases that rely on these types? Any experience you can share about using them?

If we have a proposal that supports these types, I agree that we should probably support both the equality operators defined in F&O and comparison operators that would support `ORDER BY`.

**Comment by ericprud** [comment]
@jaw111, doesn't [ORDER BY](https://www.w3.org/TR/sparql11-query/#modOrderBy)'s reliance on ['<'](https://www.w3.org/TR/sparql11-query/#op_lt) mean that we don't have to worry about 'ORDER BY'? Or are you saying we need to include tests that verify that implementations conform?

**Comment by kasei** [comment]
@ericprud I understood this comment to be about the fact that F&O doesn't explicitly enumerate a `<` operator for all the types being discussed (e.g. `gYearMonth`).

**Comment by lisp** [comment]
the deficiency in the definition of the '<' is the issue which gave rise to the suggestion.

**Comment by jaw111** [comment]
@ericprud yes, exactly as @kasei mentions, therefore implementations would need to calculate the starting instances for values with those datatypes.

@kasei I don't have any use cases that **rely** on those datatypes, but as @lisp mentions this came up recently and is roughly represented by the query I posted earlier in this thread. One could also imagine similar usage of `gMonthDay` if one were trying to identify annual patterns in a dataset.

**Comment by white-gecko** [comment]
We have a use case that is modelling data with `xsd:gDay`, `xsd:gMonth`, `xsd:gMonthDay`, `xsd:gYear`, and `xsd:gYearMonth`. We model prosopographical data (historic persons) where the sources not always specify e.g. the exact birth date so we end up with just the month and the day without knowing the year. In other cases we only know the year or year and month. In such cases we still need to compare the dates. The current workaround to perform some string concatenation and cast the result to `xsd:dateTime` for the comparison. Also see #113#issuecomment-682438571

**Comment by namedgraph** [comment]
@white-gecko we used the [Time ontology](https://www.w3.org/TR/owl-time/) for a similar use case.

**Comment by lucaturchet** [comment]
Is there any update on the issue of computing arithmetics on dates via SPARQL queries?

**Comment by afs** [comment]
Many systems implement [XPath and XQuery Functions and Operators](https://www.w3.org/TR/xpath-functions-3/) (AKA "F&O") and [XQuery and XPath Data Model](https://www.w3.org/TR/xpath-datamodel-3/), for the the datatypes that make sense in RDF and SPARQL - the XSD atomic datatypes (no XML sequences), except those relating to the structure of XML itself.

The data model and F&O are large pieces of work that carefully defines the XSD arithmetic, including the error and corner cases.

**Comment by afs** [comment]
https://github.com/w3c/sparql-12/blob/main/SEP/SEP-0002/sep-0002.md

Issue #55

**Comment by lucaturchet** [comment]
For duration there is only the operator = , but a relevant use case is that of computing the average of the durations of an event. Could you please add it?

**Comment by lisp** [comment]
please describe "the average" of the "durations" of "an event" in terms of the other xpath operators and datatypes.

**Comment by lucaturchet** [comment]
just imagine that I want to calculate the average duration of all events belonging to a certain kind (e.g., music concert) happening during a given year (e.g., 2022). This is not possible as there are no operations on xsd:duratiom with the exception of "equals".

**Comment by lisp** [comment]
i would prefer, not to imagine, but rather to consider a specific data model and the respective operator definitions.  

would you be in a position to construct a dataset which demonstrates the use case and detail the transformations which the use case requires?

**Comment by afs** [comment]
@lucaturchet -- see the section https://www.w3.org/TR/xpath-functions/#duration-subtypes The section discusses two subtypes of duration.

There are more operations "equals". There are accessors to get the components as numeric values, and - where it is meaningful - compare subtypes.

Concepts like average could only be meaningful for xs:dayTimeDuration. Even "day" is tricky because it is not a fixed length of seconds in UTC (it has leap seconds but not a fixed points) and for some communities that matters.

You can't average months - they change in length so there no fixed "half a month".

What you can do is turn the duration into integer - and the app is taking responsibility for the meaningfulness of that for the intended data.

So use an expression appropriate to the use case; there isn't a general answer. e.g. HOURS()*3600+MINUTES()*60+SECONDS() and work in numbers with HOURS etc being the natural extension of SPARQL because strict SPARQL isn't defined for xsd durations, but system may and do provide additional datatype support.

**Comment by lucaturchet** [comment]
Ok, assuming that I want to make the average of only hours, minutes and seconds, in a SPARQL query what I have to do? Can you post the exact formula to be included in after the SELECT if I have a duration expressed in xsd:duration?

**Comment by afs** [comment]
`BIND (HOURS(?x)*3600+MINUTES(?x)*60+SECONDS(?x) AS ?seconds)`, where `?x` is the duration, and you have the time interval as a number of seconds -- if the engine you are using supports the datatype - it is not required by the standard.

**Comment by lisp** [comment]
> However fn:duration-in-seconds does not exist in the F&O spec.

shouldn't it?

**Comment by afs** [comment]
That is not the point. Creating URIs in a namespace you don't control is bad semantic web practice. 

It also breaks the conventions on naming in F&O where it would include "dayTimeDuration" like the other dayTimeDuration specific operations.

**Comment by afs** [comment]
That extracts the seconds component.
`fn:seconds-from-duration("PT2H1S"^^xsd:dayTimeDuration)` is `1.0`.

**Comment by namedgraph** [comment]
And what's wrong with `xsd:dayTimeDuration(avg(?seconds)) / xs:dayTimeDuration('PT1S')`?
[op:divide-dayTimeDuration-by-dayTimeDuration](https://www.w3.org/TR/xpath-functions-31/#func-divide-dayTimeDuration-by-dayTimeDuration)

**Comment by michaelhkay** [comment]
In XPath it's "div" rather than "/". Personally, I find that manipulating durations as numbers (e.g. of years or seconds) is much more practical than manipulating them with a custom data type, and I see no reason why if number types work for lengths, weights, and voltages, they shouldn't also work for durations. I think that the reason they've been given a special type is primarily because of the interaction with dates and times, which do justify a custom data type because of calendrical complexities such as leap years and time zones. If I want to record timings for runners in a race, I'll probably choose to use numbers, not durations.

**Comment by IllidanS4** [comment]
@michaelhkay Durations are manipulated as numbers (well tuples to be precise), they are just formatted and stored in a specific way. The same goes for CSS for example, you have `16cm` etc. for length and there is no issue with that, likewise calculating `16cm + 12px` does not result in some weird quantity with both centimeters and pixels, but in a concrete scalar value. If you treat duration as a physical quantity, feel free to use seconds, but humans also use months which cannot be converted to a value in seconds.

@afs To be precise, it returns the seconds component modulo 60. `fn:seconds-from-duration("PT66S"^^xsd:dayTimeDuration)` is still `6`.

**Comment by nichtich** [comment]
It would help to update [SEP0002](https://github.com/w3c/sparql-dev/blob/main/SEP/SEP-0002/sep-0002.md) with a report on current implementations to discuss this issue. There are some test cases at https://github.com/kasei/sparql-12/tree/xsd_datetime_duration/tests/xsd_functions but not sure whether they cover everything. A good introduction is [this article](https://www.bobdc.com/blog/sparql12time/) from August 2023 but the main question "which date functions are supported by current SPARQL processors?" is not answered. To me it looks like SPARQL 1.2 specification draft is behind actual practice in support of temporal arithmetics.

**Comment by Tpt** [comment]
> There are some test cases at https://github.com/kasei/sparql-12/tree/xsd_datetime_duration/tests/xsd_functions but not sure whether they cover everything

Yes! If @kasei is fine with it, it might be nice to move the tests to this repository to allow developing a testsuite than might be then reused for SPARQL 1.2 (or more likely a follow-up version integrating new features like LATERAL...).

**Comment by afs** [comment]
> To me it looks like SPARQL 1.2 specification draft is behind actual practice in support of temporal arithmetics.

After SPARQL 1.2 is published, the working group moves into "maintenance and new features mode" (this is a W3C way of being evolve standards without a full WG cycle). Adding more functions could happen, either as a spec addition or as a related W3C Note.

"rdf-tests" can be used to capture community practice. It is a softer option that does to label a system as incomplete or "non-compliant" if there is only has partial support, which can become a political issue.

There is a lot of "[Functions and Operators](https://www.w3.org/TR/xpath-functions-3/)".

[F&O Section 8 Functions and operators on durations](https://www.w3.org/TR/xpath-functions-3/#durations)
[F&O Section 9 Functions and operators on dates and times](https://www.w3.org/TR/xpath-functions-3/#dates-times)

**Comment by nichtich** [comment]
So what's best practice in documenting support of new features? Take the list of test cases as test suite and create a conformance dashboard? An overview like https://caniuse.com/ would be perfect but a raw table such as https://github.com/karlcow/markdown-testsuite?tab=readme-ov-file#test-scripts ok as well.

**Comment by afs** [comment]
We have the machinery (EARL tests). If someone wants to set and _maintain_ such a system, it would be great. Self-reporting does tend to bit-rot quite quickly.

**Comment by Tpt** [comment]
> If someone wants to set and maintain such a system, it would be great. Self-reporting does tend to bit-rot quite quickly.

@bordercloud worked in this direction a few years ago but I am not sure how much the project is alive http://sparqlscore.com/

**Comment by namedgraph** [comment]
@Tpt @nichtich related: https://lists.w3.org/Archives/Public/public-sparql-dev/2024AprJun/0012.html
